### PR TITLE
[SRX-UKDUKT] Locks Page with Application and Environment Lock display

### PR DESCRIPTION
**Pull request by leandro-freiheit**

### CODE REVIEW

File: services/frontend-service/src/ui/Pages/Locks/LocksPage.test.tsx [hunk 1]
```diff
@@ -17,12 +17,12 @@ Copyright 2021 freiheit.com*/
...
 
-describe('EnvironmentsPage', () => {
+describe('LocksPage', () => {
     const getNode = (): JSX.Element | any => <LocksPage />;
     const getWrapper = () => render(getNode());
 
     it('Renders full app', () => {
         const { container } = getWrapper();
-        expect(container.textContent).toBe('List All Locks Here');
+        expect(container.getElementsByClassName('mdc-env-data-table')[0]).toBeTruthy();
```
Comment (medo-freiheit) [inline services/frontend-service/src/ui/Pages/Locks/LocksPage.test.tsx:26]: I think this test is useless maybe check for the two titles ' environment locks ' & ' application locks ' or maybe get rid of it

File: services/frontend-service/src/ui/components/ServiceLane/ServiceLane.tsx [hunk 2]
```diff
@@ -29,12 +30,12 @@ export const ServiceLane: React.FC<{ application: string }> = (props) => {
                     <Button
                         className="service-action service-action--prepare-undeploy"
                         label={'Prepare to delete'}
-                        icon={'delete'}
+                        icon={<DeleteWhite />}
```
Comment (tameremad) [inline services/frontend-service/src/ui/components/ServiceLane/ServiceLane.tsx:33]: just a small hint. it would make this PR much more readable if you split it up.
e.g. do all the images in another PR.
* adding the svgs
* changing the text icons and etc...
* changing the button implementation

Comment (tameremad) [reply services/frontend-service/src/ui/components/ServiceLane/ServiceLane.tsx:33]: but for now it's okay

File: services/frontend-service/src/ui/components/button/button.tsx [hunk 3]
```diff
@@ -39,7 +39,10 @@ export const Button = (props: { className?: string; label?: string; icon?: strin
             <div className="mdc-button__ripple" />
             {!!icon && (
```
Comment (tameremad) [inline services/frontend-service/src/ui/components/button/button.tsx:40]: this is two levels of complicated. you can remove the `<i/>` element and maybe one of the checks 

(if needed, add the classnames to the icon JSX)

File: services/frontend-service/src/ui/components/EnvironmentLockDisplay/EnvironmentLockDisplay.tsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                    <div className="env-lock-display-info">
+                        <Button className="lock-action service-action--delete" icon={<Delete />} />
+                    </div>
+                </div>
+            </div>
+        </div>
+    );
+};
+
+// const getFullDate = (time: number) => {
```
Comment (tameremad) [inline services/frontend-service/src/ui/components/EnvironmentLockDisplay/EnvironmentLockDisplay.tsx:46]: ?

File: services/frontend-service/src/ui/utils/store.tsx [hunk 5]
```diff
@@ -26,6 +26,70 @@ export const [_, PanicOverview] = createStore({ error: '' });
...
+                    lockInfo[1].createdBy?.name,
+                    lockInfo[1].createdBy?.email,
+                ])
+            )
+            .filter((content) => content.length !== 0)
+            .reduce((acc, val) => acc.concat(val), [])
+            .sort((a, b) => {
+                if (a !== undefined && b !== undefined) {
+                    if (a[0] !== undefined && b[0] !== undefined) {
+                        if (a[0] > b[0]) {
```
[the remark this answers is not in the document]
Comment (tameremad) [reply services/frontend-service/src/ui/utils/store.tsx:48]: Note to self: locks could be a type and `uselocks` will take one optional parameter of sortKey. and the locks will be sorted by the specified key.

File: services/frontend-service/src/ui/components/ApplicationLocksCard/ApplicationLocksCard.tsx [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+You should have received a copy of the GNU General Public License
+along with kuberpult.  If not, see <http://www.gnu.org/licenses/>.
+
+Copyright 2021 freiheit.com*/
+import { useApplicationLocks } from '../../utils/store';
+import { ApplicationLockDisplay } from '../ApplicationLockDisplay/ApplicationLockDisplay';
+
+export const ApplicationLocksCard: React.FC<{}> = (props) => {
+    const appLocks = useApplicationLocks();
+    // eslint-disable-next-line no-console
```
Comment (tameremad) [inline services/frontend-service/src/ui/components/ApplicationLocksCard/ApplicationLocksCard.tsx:22]: please remove later

File: services/frontend-service/src/ui/components/ApplicationLockDisplay/ApplicationLockDisplay.tsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
+/*This file is part of kuberpult.
+
```
Comment (tameremad) [inline services/frontend-service/src/ui/components/ApplicationLockDisplay/ApplicationLockDisplay.tsx:2]: does it make sense to combine the two `Display` components? 

all locks are similar, and they should be displayed the same way?

File: services/frontend-service/src/ui/components/ApplicationLockDisplay/ApplicationLockDisplay.scss [hunk 8]
```diff
@@ -0,0 +1,36 @@
+@use "@material/data-table/data-table";
```
Comment (tameremad) [inline services/frontend-service/src/ui/components/ApplicationLockDisplay/ApplicationLockDisplay.scss:1]: At least the CSS should be one for both components? or are there any differences here?

File: services/frontend-service/src/assets/_variables.scss [hunk 9]
```diff
@@ -22,6 +22,23 @@ $sidebar-apply-font-size: 1.5em;
...
+$locks-list-header-border-radius: 10px;
+$locks-list-header-height: 3em;
+$locks-list-header-padding: .5em;
+$locks-list-header-line-height: 2em;
+$locks-list-header-font-size: 1.1em;
+$locks-list-row-spacing: -.3em 0;
+$locks-list-row-height: 3.5em;
+$locks-list-row-margin: .5em;
+$locks-button-delete-height: 4em;
+$locks-date-outdated-color: #FF8A00;
```
Comment (tameremad) [inline services/frontend-service/src/assets/_variables.scss:34]: This style for outdated locks may not go along well with the other UI elements. 
We need to check with safi on this.

File: services/frontend-service/src/ui/components/ServiceLane/ServiceLane.tsx (hunk 2 above)
Comment (leandro-freiheit) [reply services/frontend-service/src/ui/components/ServiceLane/ServiceLane.tsx:33]: Okay, noted!

File: services/frontend-service/src/ui/Pages/Locks/LocksPage.test.tsx (hunk 1 above)
Comment (leandro-freiheit) [reply services/frontend-service/src/ui/Pages/Locks/LocksPage.test.tsx:26]: I ended up maintaining the test but now checking both the titles

**Review by medo-freiheit — COMMENTED** [review]
There are a lot of duplicated code, maybe create a reusable Card/Table component that takes `title, header fileds, locks`

File: services/frontend-service/src/ui/utils/store.tsx [hunk 10]
```diff
@@ -26,6 +26,70 @@ export const [_, PanicOverview] = createStore({ error: '' });
 export const useApplicationNames = () =>
     useOverview(({ applications }) => Object.keys(applications).sort((a, b) => a.localeCompare(b)));
 
+// return all environment locks
+export const useEnvironmentLocks = () =>
+    useOverview(({ environments }) =>
+        Object.values(environments)
+            .map((environment) =>
+                Object.entries(environment.locks).map((lockInfo) => [
```
[the remark this answers is not in the document]
Comment (leandro-freiheit) [reply services/frontend-service/src/ui/utils/store.tsx:34]: Okay, so what I ended up doing was using an interface as you said!

File: services/frontend-service/src/ui/utils/store.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (leandro-freiheit) [reply services/frontend-service/src/ui/utils/store.tsx:48]: I did a sort function that sorts by all the fields. However, I still think it can be improved.

File: services/frontend-service/src/ui/components/EnvironmentLockDisplay/EnvironmentLockDisplay.tsx (hunk 4 above)
Comment (leandro-freiheit) [reply services/frontend-service/src/ui/components/EnvironmentLockDisplay/EnvironmentLockDisplay.tsx:46]: Removed!

File: services/frontend-service/src/ui/components/ApplicationLocksCard/ApplicationLocksCard.tsx (hunk 6 above)
Comment (leandro-freiheit) [reply services/frontend-service/src/ui/components/ApplicationLocksCard/ApplicationLocksCard.tsx:22]: Removed!

File: services/frontend-service/src/ui/components/ApplicationLockDisplay/ApplicationLockDisplay.tsx (hunk 7 above)
Comment (leandro-freiheit) [reply services/frontend-service/src/ui/components/ApplicationLockDisplay/ApplicationLockDisplay.tsx:2]: Yeah, it makes sense of making into one component. Already did that!

File: services/frontend-service/src/ui/components/ApplicationLockDisplay/ApplicationLockDisplay.scss (hunk 8 above)
Comment (leandro-freiheit) [reply services/frontend-service/src/ui/components/ApplicationLockDisplay/ApplicationLockDisplay.scss:1]: No, now that only one component is used , it is more clear that only one use of data-table should be used

File: services/frontend-service/src/assets/_variables.scss (hunk 9 above)
Comment (leandro-freiheit) [reply services/frontend-service/src/assets/_variables.scss:34]: Me and @medo-freiheit already experimented with the colors. Nonetheless, I agree.

File: services/frontend-service/src/ui/components/ApplicationLocksCard/ApplicationLocksCard.tsx [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with kuberpult.  If not, see <http://www.gnu.org/licenses/>.
+
+Copyright 2021 freiheit.com*/
+import { useApplicationLocks } from '../../utils/store';
+import { ApplicationLockDisplay } from '../ApplicationLockDisplay/ApplicationLockDisplay';
+
+export const ApplicationLocksCard: React.FC<{}> = (props) => {
```
[the remark this answers is not in the document]
Comment (leandro-freiheit) [reply services/frontend-service/src/ui/components/ApplicationLocksCard/ApplicationLocksCard.tsx:20]: All fixed and now there is only one LockDisplay and one LockTable component

File: services/frontend-service/src/ui/components/button/button.tsx [hunk 12]
```diff
@@ -37,11 +37,10 @@ export const Button = (props: { className?: string; label?: string; icon?: strin
             ref={control}
             aria-label={label || ''}>
             <div className="mdc-button__ripple" />
-            {!!icon && (
-                <i className="medium material-icons mdc-list-item__graphic" aria-hidden="true">
-                    {icon}
-                </i>
-            )}
+            {icon &&
```
Comment (medo-freiheit) [inline services/frontend-service/src/ui/components/button/button.tsx:40]: Why this change here ?

File: services/frontend-service/src/ui/components/LockDisplay/LockDisplay.tsx [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+        const curTime = new Date().getTime();
+        const diffTime = curTime.valueOf() - time.valueOf();
+        const msPerDay = 1000 * 60 * 60 * 24;
+        return Math.floor(diffTime / msPerDay);
+    }
+    return -1;
+};
+
+const isOutdated = (dateAdded: Date | string | undefined): boolean => {
+    if (dateAdded !== undefined && typeof dateAdded !== 'string') {
```
Comment (medo-freiheit) [inline services/frontend-service/src/ui/components/LockDisplay/LockDisplay.tsx:64]:
```suggestion
    if (!!dateAdded && typeof dateAdded !== 'string') {
```
why not tests it like this ?

File: services/frontend-service/src/ui/components/LockDisplay/LockDisplay.tsx [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+            <div className="-lock-display__table">
+                <div className="lock-display-table">
+                    <div
+                        className={
+                            'lock-display-info date-display--' + (isOutdated(lock.date) ? 'outdated' : 'normal')
+                        }>
+                        {daysToString(calcLockAge(lock.date))}
+                    </div>
+                    <div className="lock-display-info">{lock.environment}</div>
+                    {lock.application !== undefined && <div className="lock-display-info">{lock.application}</div>}
```
Comment (medo-freiheit) [inline services/frontend-service/src/ui/components/LockDisplay/LockDisplay.tsx:34]:
```suggestion
                    { !! lock.application && <div className="lock-display-info">{lock.application}</div>}
```
is there a reason to use !== undefined ?

File: services/frontend-service/src/ui/components/LockDisplay/LockDisplay.tsx [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+                    <div className="lock-display-info">{lock.authorName}</div>
+                    <div className="lock-display-info">{lock.authorEmail}</div>
+                    <Button className="lock-display-info lock-action service-action--delete" icon={<Delete />} />
+                </div>
+            </div>
+        </div>
+    );
+};
+
+const daysToString = (days: number) => {
```
Comment (medo-freiheit) [inline services/frontend-service/src/ui/components/LockDisplay/LockDisplay.tsx:46]: There are a unit tests for these functions in the old UI, can you please use them or create new ones for those functions ?
Please move these functions to the top and export them to use them in the unit tests.

File: services/frontend-service/src/ui/components/LocksTable/LocksTable.tsx [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+                            <LockDisplay lock={lock} />
+                        ))}
+                    </tbody>
+                </table>
+            </div>
+        </div>
+    );
+};
+
+const sortLocks = (displayLocks: DisplayLock[]) =>
```
Comment (medo-freiheit) [inline services/frontend-service/src/ui/components/LocksTable/LocksTable.tsx:62]: So, are we doing the sort outside the store ? why is this better ?

File: services/frontend-service/src/ui/components/button/button.tsx (hunk 12 above)
Comment (leandro-freiheit) [reply services/frontend-service/src/ui/components/button/button.tsx:40]: It's as Tamer said. How it was before as doubling the complexity. Since the icon only needs the JSX Element, we don't need a div for it.

File: services/frontend-service/src/ui/components/LockDisplay/LockDisplay.tsx (hunk 13 above)
Comment (leandro-freiheit) [reply services/frontend-service/src/ui/components/LockDisplay/LockDisplay.tsx:64]: I did not remember the `!!` operator and since I had seen this syntax in the project, I followed that. But going to fix it!

File: services/frontend-service/src/ui/components/LocksTable/LocksTable.tsx (hunk 16 above)
Comment (leandro-freiheit) [reply services/frontend-service/src/ui/components/LocksTable/LocksTable.tsx:62]: Fixed!

File: services/frontend-service/package.json [hunk 17]
```diff
@@ -81,5 +81,5 @@
       "last 1 safari version"
     ]
   },
-  "proxy": "http://kuberpult-frontend-service:8081"
+  "proxy": "http://localhost:8081"
```
Comment (medo-freiheit) [inline services/frontend-service/package.json:84]: @leandro-freiheit please return this to 
```suggestion
  "proxy": "http://kuberpult-frontend-service:8081"
```

File: services/frontend-service/src/ui/utils/store.tsx [hunk 18]
```diff
@@ -26,6 +26,73 @@ export const [_, PanicOverview] = createStore({ error: '' });
...
+                    )
+                    .filter((displayLock) => displayLock.length !== 0)
+            )
+            .filter((displayLock) => displayLock.length !== 0)
+            .map((displayLock) => displayLock[0])
+            .flat();
+        return sortLocks(locks);
+    });
+
+const sortLocks = (displayLocks: DisplayLock[]) =>
```
Comment (medo-freiheit) [inline services/frontend-service/src/ui/utils/store.tsx:79]: I'm really not feeling good about this function, it'stoo complex sorting, and it only sorts in ascending order, what about the descending order ?

Comment (leandro-freiheit) [reply services/frontend-service/src/ui/utils/store.tsx:79]: I agree with you, and It is already implemented. I allow that the locks to be sorted by ascending or descending order (first it orders by date, then environment, then application, and so on).
